[RFC] Fix issue #7429.

glacambre opened the pull request:
Link to the issue: https://github.com/neovim/neovim/issues/7429
The problem was that after a help window was closed, a window was selected and
its autocommands triggered. After that, restore_snapshot was called and the
focused window changed, confusing the user.
This commit makes sure restore_snapshot is called before any window is selected
or autocommand triggered.

glacambre commented:
I think I made a mistake - even though the current code fixes the issue it's probably wrong, since close_curwin is used before it is set.
Going back to working on it.

glacambre commented:
I fixed my mistake, I think this pull requests fixes #7429 and is correct.

justinmk commented:
Nice catch, this seems reasonable. CI failures are invalid.

Can you add a test to `test/functional/autocmd/bufenter_spec.lua` ? You can run it locally like this:

    TEST_FILE=test/functional/autocmd/bufenter_spec.lua make functionaltest

glacambre commented:
This last commit adds a test that succeeds on my branch but fails on master, as expected.

glacambre commented:
This last commit adds a test that succeeds on my branch but fails on master, as expected.

glacambre commented:
Don't merge this PR yet, I'm starting to think I did something wrong again (my PR doesn't bring the cursor back to where it should be).

chrisbra commented:
what exactly is the bug in Vim? Can you give reproducible example please?

glacambre commented:
@chrisbra After opening any file A, open a new buffer B in a horizontal split, move the cursor to buffer A, open a help window, move the help window to the right, create an autocommand specific to buffer B, wipe the help window. Basically this:
```vimscript
split B
wincmd j
help help
wincmd L
autocmd BufEnter B let g:triggered_bufenter = 1
bw
```
When the help window is closed, the cursor is first moved to buffer B, the autocommand is triggered and then the cursor is moved to buffer A. This can be very confusing, since the only thing the user sees is the cursor moving to buffer A and buffer B's autocmd being triggered.

chrisbra commented:
ah, I see. Can reproduce. CC @brammool please check referenced patch.

glacambre commented:
The current patch is wrong and breaks `:help`'s current behaviour which is to put the cursor back to where it was when the help window is closed.
However I don't have a lot of time to spend on it so progress might be slow, if you'd rather fix it yourself feel free to do so :).

brammool commented:
Moving restore_snapshot() is not going to work.  Better get the new curwin from the snapshot, instead of guessing what the next curwin will be.  Thus after:
    wp = win_free_mem(win, &dir, NULL);

Commit:
bufenter_spec.lua: test for BufEnter+:help interaction

Commit:
window.c: Fix issue #7429

Link to the issue: https://github.com/neovim/neovim/issues/7429
Problem: after a help window was closed, a window was selected and its
autocommands triggered. After that, restore_snapshot was called and the focused
window changed, confusing the user.

Solution: Add a function get_snapshot_focus() that returns the window
that holds the cursor in a snapshot. This function is used in win_close
to make sure the right window is selected before any autocommand is
triggered.

justinmk commented:
@glacambre Is this ready ?

glacambre commented:
I think it is ready, if there's anything that needs to be changed please let me know.

justinmk commented on src/nvim/window.c:
```diff
@@ -5421,6 +5430,29 @@ static win_T *restore_snapshot_rec(frame_T *sn, frame_T *fr)
+}
+
+/* Get the focused window (the one holding the cursor) of the snapshot */
+win_T *get_snapshot_focus(int idx)
```
can be `static`

justinmk commented on src/nvim/window.c:
```diff
@@ -5421,6 +5430,29 @@ static win_T *restore_snapshot_rec(frame_T *sn, frame_T *fr)
+{
+  if (curtab->tp_snapshot[idx] == NULL)
+    return NULL;
+  return get_snapshot_focus_rec(curtab->tp_snapshot[idx]);
```
Since it's not recursive, there's no need for a separate function (and the `_rec` suffix).

justinmk commented:
Merged after fixing lint.
